src/main/java/org/testng/internal/Invoker.java

```diff
@@ -1123,153 +907,142 @@ private ParameterBag createParameters(ITestNGMethod testMethod,
-          } else {
-            tr.setStatus(ITestResult.SKIP);
-            m_notifier.addSkippedTest(testMethod, tr);
-          }
-          runTestListeners(tr);
-          result.add(tr);
-          continue;
+    final ITestNGMethod[] beforeMethods = TestNgMethodUtils.filterBeforeTestMethods(testClass, CAN_RUN_FROM_CLASS);
+    final ITestNGMethod[] afterMethods = TestNgMethodUtils.filterAfterTestMethods(testClass, CAN_RUN_FROM_CLASS);
+    while (invocationCount-- > 0) {
+      long start = System.currentTimeMillis();
+
+      Map<String, String> allParameterNames = Maps.newHashMap();
+      ParameterHandler handler = new ParameterHandler(m_annotationFinder, m_dataproviderListeners);
+
+      ParameterBag bag = handler.createParameters(testMethod, parameters, allParameterNames, testContext, instance);
+
+      if (bag.hasErrors()) {
+        final ITestResult tr = bag.errorResult;
+        Throwable throwable = tr.getThrowable();
+        if (throwable instanceof TestNGException) {
+          tr.setStatus(ITestResult.FAILURE);
+          m_notifier.addFailedTest(testMethod, tr);
+        } else {
+          tr.setStatus(ITestResult.SKIP);
+          m_notifier.addSkippedTest(testMethod, tr);
         }
+        runTestListeners(tr);
+        result.add(tr);
+        continue;
+      }
 
-        Iterator<Object[]> allParameterValues = bag.parameterHolder.parameters;
-        int parametersIndex = 0;
-
-        try {
-          if (bag.parameterHolder.origin == ParameterOrigin.ORIGIN_DATA_PROVIDER &&
-              bag.parameterHolder.dataProviderHolder.isParallel()) {
-            List<TestMethodWithDataProviderMethodWorker> workers = Lists.newArrayList();
-            while (allParameterValues.hasNext()) {
-              Object[] next = allParameterValues.next();
-              if (next == null) {
-                // skipped value
-                parametersIndex++;
-                continue;
-              }
-              Object[] parameterValues = injectParameters(next,
-                  testMethod.getConstructorOrMethod().getMethod(), testContext, null /* test result */);
-
-              TestMethodWithDataProviderMethodWorker w =
-                new TestMethodWithDataProviderMethodWorker(this,
-                    testMethod, parametersIndex,
-                    parameterValues, instance, suite, parameters, testClass,
-                    beforeMethods, afterMethods, groupMethods,
-                    expectedExceptionHolder, testContext, m_skipFailedInvocationCounts,
-                    invocationCount, failure.count, m_notifier);
-              workers.add(w);
-              // testng387: increment the param index in the bag.
+      Iterator<Object[]> allParameterValues = bag.parameterHolder.parameters;
+      int parametersIndex = 0;
+
+      try {
+
+        if (bag.runInParallel()) {
+          List<TestMethodWithDataProviderMethodWorker> workers = Lists.newArrayList();
+          while (allParameterValues.hasNext()) {
```

Comment:
Is a foreach possible here?

Reply:
`allParameterValues` is of type `Iterator<Object[]>`. Can't use `foreach`

Reply:
ok, it will be too big to change it by something more compliant (like collection or Iterable)